Allow dataflow command line programs to run on a Hadoop/Spark cluster.

tomwhite opened the pull request:
This is achieved by i) using HadoopBAM to read BAM files from HDFS, and
ii) adding a shadowJar target to build app JAR suitable for running with Spark.

Note that this uses Spark's spark-submit script to run on the cluster - this is to avoid including the Spark classes in the hellbender JAR itself, so that it can run on different versions of Spark. We might consider making a runnable JAR at some point in the future though.

I tested this on a standalone cluster (running with Java 8) as follows:

```bash
export SPARK_HOME=~/sw/spark-1.3.1-bin-hadoop2.6/ 
PATH=$PATH:$SPARK_HOME/bin 
$SPARK_HOME/sbin/start-all.sh

export HADOOP_HOME=~/sw/hadoop-2.7.0
export HADOOP_CONF_DIR=$HADOOP_HOME/etc/hadoop.pseudo/
export PATH=$PATH:$HADOOP_HOME/bin:$HADOOP_HOME/sbin
(cd $HADOOP_HOME; start-hdfs2)
hadoop fs -mkdir -p /user/$USER 
hadoop fs -put ~/workspace/hellbender/src/test/resources/org/broadinstitute/hellbender/tools/flag_stat.bam flag_stat.bam
hadoop fs -rmr out

SPARK_MASTER=spark://localhost:7077
spark-submit \
  --class org.broadinstitute.hellbender.Main \
  --master $SPARK_MASTER \
  build/libs/hellbender-GATK.4.alpha-*-all.jar CountBasesDataflow \
    --input hdfs://localhost/user/$USER/flag_stat.bam \
    --output hdfs://localhost/user/$USER/out/countbases \
    --runner SPARK \
    --sparkMaster $SPARK_MASTER
hadoop fs -getmerge out out
cat out
```

lbergelson commented on build.gradle:
```diff
@@ -88,17 +91,19 @@ dependencies {
     testCompile 'org.hamcrest:hamcrest-all:1.3'
     testCompile 'junit:junit:4.12'
-
+    testCompile('org.apache.spark:spark-core_2.10:1.3.1') {
```
Much nicer to have this as a test dependency.

lbergelson commented on build.gradle:
```diff
@@ -193,6 +198,17 @@ task fatJar(type: Jar) {
     with jar
 }
 
+shadowJar {
```
Could you explain why this is needed?  I assume it's to prevent conflicts with hadoop dependencies in some way, but I don't really know what it's doing exactly.

tomwhite replied:
It's so that guava can be relocated to avoid conflicts with Hadoop's version. (The dataflow SDK also relocates guava in the same way BTW.) The result is basically the same as the fatJar, so fatJar could probably be replaced with shadowJar at some point.

lbergelson replied:
does it produce a complete jar with all the dependencies bundled in it?

tomwhite replied:
Yes, all the non-test dependencies.

lbergelson replied:
great, could you delete fatJar as part of this then?  No need to have both if the shadow jar includes it all.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/DataflowCommandLineProgram.java:
```diff
@@ -89,19 +95,28 @@ public static String getRunnerTypeName(String runnerSimpleName) {
+            final SparkPipelineOptions options = PipelineOptionsFactory.as(SparkPipelineOptions.class);
+            options.setRunner(this.runnerType.runner);
+            options.setSparkMaster(sparkMaster);
+            pipelineOptions = options;
```
could you extract this into a method like `createPipelineOptions`.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/DataflowUtils.java:
```diff
@@ -67,6 +73,50 @@ public void processElement(ProcessContext c) {
+     * @return a <code>PCollection<Read></code> with all the reads that overlap the
+     * given intervals in the BAM files
+     */
+    public static PCollection<Read> getReadsFromHadoopBams(final Pipeline pipeline, final List<SimpleInterval> intervals, final String path) {
```
We need to have tests for this.  I'm not totally sure how we should go about doing that because we don't have a hadoop filesystem set up in travis. 

 Is there an easy way to serve files as hdfs to test things?

lbergelson replied:
Does this actually read multiple bam files if you provide directory with multiple files as the path?  

If that is the case could you change it so that it accepts a list of paths instead and only reads from the explicitly specified bams?  

If it only reads from 1 at a time, could you change the name/ documentation.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/DataflowUtils.java:
```diff
@@ -67,6 +73,50 @@ public void processElement(ProcessContext c) {
+    /**
+     * Tests if a given record overlaps any interval in a collection.
+     */
+    private static boolean overlaps(SAMRecord record, List<SimpleInterval> intervals) {
```
This is an unfortunate omission from the SimpleInterval API.   It's come up in a number of places that it's needed.  

The interval lists can be very large (100k+ intervals), so iterating over each interval is a bit scary in terms of performance.  We should adapt the IntervalTree from htsjdk so that we can do this operation efficiently, but that's outside of the scope of this pull request.

lbergelson replied:
I created a ticket for this https://github.com/broadinstitute/hellbender/issues/559

lbergelson replied:
for now could you rewrite this method in terms of SimpleInterval.overlaps(Locatable l)  and add a todo to update it when #559 is resolved?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/ReadsSource.java:
```diff
@@ -46,6 +50,7 @@ public ReadsSource(String bam, Pipeline p){
         this.pipeline = p;
 
         cloudStorageUrl = BucketUtils.isCloudStorageUrl(bam);
+        hadoopUrl = BucketUtils.isHadoopUrl(bam);
```
More evidence that we need definitely need #477 .

lbergelson commented:
@tomwhite  Review complete.  Thanks for this!   

I'm at a bit of a loss for how to test the hadoop bam reading.  I think we should have a unit test for it, but I don't know the right way to do it.  Is there an easy way to spin up something that serves local files as hdfs?  Ideally we could have an hdfs layer running on travis that we could query files from in order to test. 

At a minimum we need unit tests testing that we can actually read an hdfs file. Even if they have to be stuck in their own hdfs group and excluded from the travis set, we need some thing that can be run to verify that it works.

tomwhite commented:
Thanks for all the comments, @lbergelson. For the Hadoop BAM reading, I've added a test for `getReadsFromHadoopBam` that uses a local file rather than HDFS, but still exercises the Hadoop code path (this is quite a common technique in testing with Hadoop).

I've also addressed your other comments.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java:
```diff
@@ -186,17 +186,17 @@ public final int size() {
      * @param other interval to check
      * @return true if this interval overlaps other, otherwise false
      */
-    public final boolean overlaps( final SimpleInterval other ) {
```
Thanks for generalizing these methods. I had miss-remembered them as already working on `Locatable` directly.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java:
```diff
@@ -186,17 +186,17 @@ public final int size() {
-    public final boolean overlaps( final SimpleInterval other ) {
-        if ( other == null ) {
+    public final boolean overlaps( final Locatable other ) {
+        if (other == null) {
```
Could you also include a check if `other.getContig() == null` as well.  It's a valid case for general `Locatable`.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java:
```diff
@@ -205,11 +205,11 @@ public final boolean overlaps( final SimpleInterval other ) {
      */
-    public final boolean contains( final SimpleInterval other ) {
+    public final boolean contains( final Locatable other ) {
         if ( other == null ) {
```
check for null contig here too

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/DataflowUtils.java:
```diff
@@ -84,6 +91,56 @@ public void processElement(ProcessContext c) {
+     */
+    private static boolean overlaps(SAMRecord record, List<SimpleInterval> intervals) {
+        if (intervals == null || intervals.isEmpty()) {
+            return true;
```
Shouldn't empty intervals not overlap?  

In most places in Hellbender we treat an empty interval list as unspecified and therefore including the whole genome.

tomwhite replied:
Perhaps I'm missing something, but isn't that what this code does? If no intervals are specified (indicated by null, or an empty list) then they overlap, and therefore include the whole genome.

lbergelson replied:
Oh, ack.  You're right.  I got twisted around about the result.  You're right.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/DataflowUtils.java:
```diff
@@ -84,6 +91,56 @@ public void processElement(ProcessContext c) {
+
+    /**
+     * Tests if a given record overlaps any interval in a collection.
+     */
```
Could you leave a TODO here referring to issue #559.  That way it will be searchable when someone actually addresses that ticket.

lbergelson commented:
Thanks for the update Tom.  Looks good.  Could you address the few new comments and then everything is :+1:

tomwhite commented:
Thanks again for the review, Louis. I've actually left fatJar in, as I was seeing some VerifyErrors from the relocated Guava classes when running the jar. They may be harmless as the job on the Spark cluster works, but I thought it would be best to keep two jars for the moment (the Spark one has a 'spark' suffix).

I've addressed the other comments, save one - not sure if that needs any changes.

lbergelson commented:
Thanks for the update Tom. Looks good.   Waiting on :+1:  from travis and then good to go.
